import { HttpHeaders, HttpErrorResponse } from "@angular/common/http";
import { throwError } from "rxjs";
import { environment } from "src/environments/environment";

export abstract class BaseService {

    protected UrlServiceV1: string = environment.apiUrlv1;

    protected ObterHeaderJson() {
        return {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
    }

    protected extractRetorno(response: any) {
        console.log(response.retorno);
        return response.retorno || {};
    }

    protected extractRegistros(response: any) {
        console.log(response.retorno.registros);
        return response.retorno.registros || {};
    }

    protected serviceError(response: Response | any) {
        let customError: string[] = [];
        let customResponse = { error: { mensagens: [] }}

        if (response instanceof HttpErrorResponse) {

            if (response.statusText === "Unknown Error") {
                customError.push("Ocorreu um erro desconhecido");
                response.error.mensagens = customError;
            }
        }
        if (response.status === 500) {
            customError.push("Ocorreu um erro no processamento, tente novamente mais tarde ou contate o nosso suporte.");
            
            // Erros do tipo 500 nÃ£o possuem uma lista de erros
            // A lista de erros do HttpErrorResponse Ã© readonly                
            customResponse.error.mensagens = customError;
            return throwError(customResponse);
        }
        
        console.error(response);
        return throwError(() => response);
    }
}